### PR TITLE
rsnapshot service: Avoid package rebuild, create+use /etc/rsnapshot.conf

### PR DESCRIPTION
**Pull request by Shados**
Doing things this way seems a lot more sensible to me - it avoids an unnecessary package rebuild, and you can use commands like `rsnapshot configtest` and `rsnapshot du` without having to refer to `/etc/crontab` to determine the correct `rsnapshot` binary to use to do so...

### CODE REVIEW

**Comment by oxij** [comment]
Personally, I (ab)use the current behavior: the daemon runs with it's
own configuration (common to many machines) and /etc/rsnapshot.conf
describes a different one, private to the host.

So I'd like to have a choice.

**Comment by aristidb** [comment]
Rebuilding the rsnapshot package is pretty trivial, as it's just a Perl script.

Furthermore, rsnapshot du should work without referring to the crontab, as it is also added to environment.systemPackages. If it doesn't work for you, it might be because you added an unpatched rsnapshot to your user profile?

To be honest, I forgot why I exactly I did it this way, though. Maybe because I disliked the impurity of /etc?

**Comment by Shados** [comment]
@oxij as this patch still respects the `package` attribute, you can just override that if you really want a separate package built. Although, you could just use the `-c cfgfile` argument to the `rsnapshot` command if you want to use a non-standard location.

@aristidb Not sure why it wasn't working for me, I don't have the default `rsnapshot` package added anywhere else. Rebuilding it is definitely trivial, but it also seems really pointless - if avoiding `/etc/` is beneficial, we can pass in the `/nix/store` config file as an argument in the crontab line. This would also be more in line with how other NixOS service implementations work, as they generally either use the standard config file location or are passed the location on the command line.

**Comment by layus** [comment]
This is indeed not the way to go because that path may be altered or overridden.
Generating and passing the `/nix/store/...-configuration` file as a `-c` argument is the most used pattern in nixos and should be used here.

Overriding the derivation argument is indeed not the way to go, simply because it requires to maintain the possibility to set that argument in the derivation itself. The patch was introduced in #1045 to ensure that rsnapshot uses /etc/rsnapshot.conf by default.

I believe that rsnapshot now uses /etc/rsnapshot.conf by default, and has the --sysconfdir option.

@Shados, would you mind updating your pull-request to use the `-c` option to rsnapshot to pass the configuration file instead of environment.etc."rsnapshot.conf"
You could also cleanup the rsnapshot derivation to remove the uneeded patches and stuff (revert #1045), but that is not mandatory.

That would be merged immediately :-).

**Comment by joachifm** [comment]
@Shados ping

**Comment by Shados** [comment]
@layus @oxij Sorry for the lateness on my part, have been busy with a new job :).

I've rewritten the PR into three cleanly delineated commits doing the following:
1: Change the rsnapshot module to pass the config file explicitly, avoiding the package rebuild
2: Modify the rsnapshot derivation to remove the config file argument, refactor the one-line patch to a `substituteInPlace`
3: Add a default-on option to the module that causes the `rsnapshot` pkg to be added to the system packages and the module-built rsnapshot config file to be symlinked to `/etc/rsnapshot.conf`

I feel doing it this way is probably the best solution, as the `rsnapshot` command is technically both an automated/service-usage command and a directly user-facing program. In NixOS, we explicitly pass config files for the former, but we do still build global `/etc/` files for the latter (e.g. `/etc/atoprc`).

I've defaulted the new option to enabled as it avoids confusion/surprise for new NixOS users, and it's trivial enough to disable if you want to do something non-standard. Disabling it should work fine for @oxij's use-case as I understand it.

File: pkgs/tools/backup/rsnapshot/default.nix [hunk 1]
```diff
@@ -28,7 +13,9 @@ stdenv.mkDerivation rec {
   patchPhase = ''
     substituteInPlace "Makefile.in" --replace \
       "/usr/bin/pod2man" "${perl}/bin/pod2man"
-    patch -p0 <${patch}
+    substituteInPlace "rsnapshot-program.pl" --replace \
+      '$config_file = $default_config_file;' \
+      '$config_file = "/etc/rsnapshot.conf";'
   '';
```
Comment (layus) [inline pkgs/tools/backup/rsnapshot/default.nix]: It seems to me that `configureFlags = [ "--sysconfdir=/etc" ]` would have the same effect without touching the internals of rsnapshot.

After some (long) investigations, the correct way to build rsnapshot is to use
```nix
{
  # ...
  configureFlags = [ "--sysconfdir=/etc --prefix=/" ];
  makeFlags = [ "DESTDIR=$(out)" ];
}
```
See for example https://git.archlinux.org/svntogit/community.git/tree/trunk/PKGBUILD?h=packages/rsnapshot.

For me, the build works with 
```diff
@@ -1,2 +1 @@
-{ fetchurl, stdenv, writeText, perl, openssh, rsync, logger,
-  configFile ? "/etc/rsnapshot.conf" }:
+{ fetchurl, stdenv, writeText, perl, openssh, rsync, logger }:
@@ -4,14 +2,0 @@
-let patch = writeText "rsnapshot-config.patch" ''
---- rsnapshot-program.pl       2013-10-05 20:31:08.715991442 +0200
-+++ rsnapshot-program.pl       2013-10-05 20:31:42.496193633 +0200
-@@ -383,7 +383,7 @@
-       }
-       
-       # set global variable
--      $config_file = $default_config_file;
-+      $config_file = '${configFile}';
- }
- 
- # accepts no args
-'';
-in
@@ -31 +15,0 @@ stdenv.mkDerivation rec {
-    patch -p0 <${patch}
@@ -33,0 +18,4 @@ stdenv.mkDerivation rec {
+  configureFlags = [ "--sysconfdir=/etc --prefix=/" ];
+
+  makeFlags = [ "DESTDIR=$(out)" ];
+
```

**Comment by Shados** [comment]
@layus: Mm, I'd assumed that something like that wouldn't work, given how it'd been done in the derivation originally, but clearly that was a mistake. Thanks for your investigative work; I've changed the relevant commit to match and tested it.

**Comment by joachifm** [comment]
Is this ready?

**Comment by Shados** [comment]
As far as I know, it should be.

**Comment by joachifm** [comment]
Okay. Thank you

**Comment by layus** [comment]
Thanks for writing these patches, it seem you got all the requirements right :-). Congratulations!

**Comment by oxij** [comment]
That's nice, thanks!

**Pull request merged**